controllers/cloud.redhat.com/providers/database/single.go

```diff
@@ -0,0 +1,302 @@
+		return dbCfg, err
+	}
+
+	svc := &core.Service{}
+	svc.ObjectMeta.OwnerReferences = ownerrefs
+	if err := db.Cache.Create(SingleDBService, nn, svc); err != nil {
+		return dbCfg, err
+	}
+
+	provutils.MakeLocalDBService(svc, nn, db.Env, labels)
+
+	if err = db.Cache.Update(SingleDBService, svc); err != nil {
+		return dbCfg, err
+	}
+
+	return dbCfg, nil
+}
+
+func (db *singleDbProvider) createDatabasePVC() error {
+	nn := db.DBNamespacedName()
+
+	pvc := &core.PersistentVolumeClaim{}
+	if err := db.Cache.Create(SingleDBPVC, nn, pvc); err != nil {
+		return err
+	}
+
+	// TODO handle volume capacity
+	capacity := sizing.GetVolCapacityForSize(sizingconfig.SizeLarge)
+	provutils.MakeLocalDBPVC(pvc, nn, db.Env, capacity)
+	if err := db.Cache.Update(SingleDBPVC, pvc); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (db *singleDbProvider) Provide(app *crd.ClowdApp) error {
+	if app.Spec.Database.SharedDBAppName != "" {
+		return db.processSharedDB(app)
+	} else if app.Spec.Database.Name != "" {
+		return db.provideAppDB(app)
+	}
+	return nil
+}
+
+func (db *singleDbProvider) provideAppDB(app *crd.ClowdApp) error {
+	dbnn := db.DBNamespacedName()
+	dbCfg := &config.DatabaseConfig{}
+
+	if err := provutils.ReadDbConfigFromSecret(db.Provider, SingleDBSecret, dbCfg, dbnn); err != nil {
+		return err
+	}
+
+	// Create database config and secret for the app,
+	// without the admin credentials.
+	appnn := types.NamespacedName{
+		Name:      fmt.Sprintf("%v-db", app.Name),
+		Namespace: app.Namespace,
+	}
+	appDbCfg, err := db.createOrReadDbConfig(app, appnn, app.Name, false)
+	if err != nil {
+		return err
+	}
+	db.Config.Database = appDbCfg
+
+	// reconcile access
+	if err := db.reconcileDBAppAccess(dbCfg, appDbCfg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (db *singleDbProvider) processSharedDB(app *crd.ClowdApp) error {
+	refApp, err := crd.GetAppForDBInSameEnv(db.Ctx, db.Client, app)
+	if err != nil {
+		return err
+	}
+
+	nn := types.NamespacedName{
+		Name:      fmt.Sprintf("%v-db", refApp.Name),
+		Namespace: refApp.Namespace,
+	}
+	dbCfg := &config.DatabaseConfig{}
+	if err := provutils.ReadDbConfigFromSecret(db.Provider, SingleDBSecret, dbCfg, nn); err != nil {
+		return errors.Wrap(fmt.Sprintf("sharedDBApp %s", refApp.Name), err)
+	}
+
+	db.Config.Database = dbCfg
+	return nil
+}
+
+func (db *singleDbProvider) Hostname() string {
+	nn := db.DBNamespacedName()
+	return fmt.Sprintf("%v.%v.svc", nn.Name, nn.Namespace)
+}
+
+func (db *singleDbProvider) createOrReadDbConfig(obj obj.ClowdObject, nn types.NamespacedName, username string, setAdmin bool) (cfg *config.DatabaseConfig, err error) {
+	cfg = &config.DatabaseConfig{}
+	password, err := utils.RandPassword(16, provutils.RCharSet)
+	if err != nil {
+		return cfg, errors.Wrap("password generate failed", err)
+	}
+
+	var pgPassword string
+	if setAdmin {
+		pgPassword, err = utils.RandPassword(16, provutils.RCharSet)
```

Comment:
My initial thought was to disable admin user for apps, however some apps might still require this to do some setups (i.e. create additional users). @psav any suggestions how we should approach this situation? I guess, one of the option is to update those services needing such privileges, or look at the privileges case-by-case.

Reply:
It's tricky because ephemeral usage often relies on them needing to do more stuff than usual, but in prod they will just get the credentials which are appropriate

Reply:
Yeah. I want to test those apps to see how they would work with this limited access... AFAIK Advisor, has dropped or is dropping the HLL extension.

Reply:
We plan to remove use of the HLL extension next week (17th Feb 2025).

Reply:
That's the only extension I'm aware of in Advisor.